Refactor path storage

peterschrammel opened the pull request:
Removes 70 lines and a lot of complexity from the single-path symex checker.

Fixes #4071

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Put initial state into path_storage

Allows us to perform the actual pathwise symex work
in a uniform way.
Required making path_storaget::push accept a single state.

peterschrammel committed:
Expose equation post-processing

Will allow us to remove perform_symex utilities later

peterschrammel committed:
Simplify single-path symex checker

Instead of handling the first symex run differently:
Push initial state into the worklist,
then iterate symex until done.

peterschrammel committed:
Inline perform_symex utility functions

Only used once resp. obsolete.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 792d854c2eecf79655d33ea934e6f4280d2499ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100012351

tautschnig commented on src/goto-symex/path_storage.cpp:
```diff
@@ -29,12 +29,9 @@ path_storaget::patht &path_lifot::private_peek()
-void path_lifot::push(
-  const path_storaget::patht &next_instruction,
-  const path_storaget::patht &jump_target)
+void path_lifot::push(const path_storaget::patht &path)
```
I don't think we actually need these overrides anymore.

peterschrammel replied:
The interface doesn't define the storage used.

tautschnig commented on src/goto-symex/path_storage.h:
```diff
@@ -68,16 +68,8 @@ class path_storaget
-  virtual void
-  push(const patht &next_instruction, const patht &jump_target) = 0;
+  /// \brief Add a path to resume to the storage
+  virtual void push(const patht &) = 0;
```
I'd suggest to not make this one abstract - all current implementations are the same, so we might as well provide a default implementation. Future deriving classes are still free to override.

peterschrammel replied:
They would also have to inherit unused internal storage.

tautschnig replied:
Ah, sorry for missing that.

tautschnig commented on src/goto-checker/single_path_symex_only_checker.cpp:
```diff
@@ -37,29 +37,24 @@ operator()(propertiest &properties)
+  auto get_goto_function =
+    [this](const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
```
Could this also be refactored? Code copies are undesirable...

peterschrammel replied:
Yes... unfortunately, C++ lambdas are so clumsy that a one-line lambda feels like code duplication...

tautschnig replied:
Shouldn't we really have an interface that doesn't require passing this parameter? I believe that in fact we already do, in `goto_symext::symex_with_state`, which has a variant with and without that function. So maybe we could also do that for those other functions?

peterschrammel replied:
> variant without that function

That one should actually be removed.

tautschnig replied:
Ok, then maybe that could happen here as well? I dislike code copies, but I even more dislike almost-code-copies which we currently have with `get_function_from_goto_functions`.

peterschrammel replied:
Check the changes, please.

tautschnig commented on src/goto-checker/multi_path_symex_checker.cpp:
```diff
@@ -41,14 +41,13 @@ operator()(propertiest &properties)
+    auto get_goto_function =
+      [this](const irep_idt &id) -> const goto_functionst::goto_functiont & {
+      return goto_model.get_goto_function(id);
+    };
```
The third copy?

tautschnig commented on src/goto-checker/multi_path_symex_only_checker.cpp:
```diff
@@ -39,14 +39,13 @@ multi_path_symex_only_checkert::multi_path_symex_only_checkert(
+  auto get_goto_function =
+    [this](const irep_idt &id) -> const goto_functionst::goto_functiont & {
+    return goto_model.get_goto_function(id);
+  };
```
Copy nr#4

peterschrammel committed:
Remove obsolete variant of symex_with_state

This variant did not support lazy loading
and was only used by scratch_program,
which can as well use the more general variant.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5281fa654b9674517fa6b9aef67c4060de6a77d8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100093034

chrisr-diffblue reviewed (approved):
Looks good to me.
